openedx/core/djangoapps/content/course_structures/api/v0/api.py

```diff
@@ -61,14 +63,39 @@ def course_structure(course_key):
     Raises:
         CourseStructureNotAvailableError, CourseNotFoundError
     """
-    course = _retrieve_course(course_key)
-    try:
-        requested_course_structure = models.CourseStructure.objects.get(course_id=course.id)
-        return serializers.CourseStructureSerializer(requested_course_structure.structure).data
-    except models.CourseStructure.DoesNotExist:
-        # If we don't have data stored, generate it and return an error.
-        tasks.update_course_structure.delay(unicode(course_key))
-        raise CourseStructureNotAvailableError
+    modified_timestamp = models.CourseStructure.objects.filter(course_id=course_key).values('modified')
+
+    if modified_timestamp.exists():
+        cache_key = 'openedx.content.course_structures.api.v0.api.course_structure.{}.{}.{}'.format(
+            course_key, modified_timestamp[0]['modified'], '_'.join(block_types or [])
+        )
+        data = cache.get(cache_key)  # pylint: disable=maybe-no-member
+        if data:
+            return data
+
+        course = _retrieve_course(course_key)
+        try:
+            course_structure = models.CourseStructure.objects.get(course_id=course.id)
+            structure = course_structure.structure
+
+            if block_types is not None:
+                blocks = course_structure.ordered_blocks
+                required_blocks = OrderedDict()
+                for usage_id, block_data in blocks.iteritems():
+                    if block_data['block_type'] in block_types:
+                        required_blocks[usage_id] = block_data
+
+                structure['blocks'] = required_blocks
+
+            data = serializers.CourseStructureSerializer(structure).data
+            cache.set(cache_key, data)  # pylint: disable=maybe-no-member
+            return data
+        except models.CourseStructure.DoesNotExist:
+            pass
+
+    # If we don't have data stored, generate it and return an error.
+    tasks.update_course_structure.delay(unicode(course_key))
+    raise CourseStructureNotAvailableError
```

Comment:
The control flow looks a bit odd here, since I would normally expect the exception handling to be around a much smaller block (i.e. just around the model access, not the whole thing). But in terms of the general logic within the function, yeah, that looks fine, ty.